### ISSUE TITLE
'NoneType' object has no attribute '_source_type'

### ISSUE THREAD

**Issue by lazka** [python-poetry/poetry#3362]
Poetry 1.1.4 on Debian sid and Ubuntu 20.10:

* `poetry init`
* `poetry add pyyaml==5.3.1 -vvv`

```python
  AttributeError

  'NoneType' object has no attribute '_source_type'

  at ~/.local/lib/python3.8/site-packages/poetry/inspection/info.py:548 in from_directory
      544│                         raise
      545│ 
      546│                     # we discovered PkgInfo but no requirements were listed
      547│ 
    → 548│         info._source_type = "directory"
      549│         info._source_url = path.as_posix()
      550│ 
      551│         return info
      552│ 
```

**Comment by lazka** [comment]
Downgrading to 1.0.10 makes things work again

**Comment by finswimmer** [comment]
Hello @lazka,

I'm unable to reproduce it. So let's start with the default questions: How have you installed poetry? Can you show the output of `poetry config --list`?

fin swimmer

**Comment by lazka** [comment]
```shell
> poetry config --list
cache-dir = "/home/lazka/.cache/pypoetry"
experimental.new-installer = true
installer.parallel = true
virtualenvs.create = true
virtualenvs.in-project = null
virtualenvs.path = "{cache-dir}/virtualenvs"  # /home/lazka/.cache/pypoetry/virtualenvs
```

**Comment by lazka** [comment]
@finswimmer try clearing your poetry cache. If pyyaml got installed with an older poetry before then it also works with the newer one because it doesn't download the sdist.

**Comment by lazka** [comment]
It seems to fail because requires_dist is None, but the package has no dependencies, so that seems.. expected?

**Comment by lazka** [comment]
I've created #3381 as a start

**Comment by lazka** [comment]
ok, turns out updating all dependencies of poetry fixes it somehow

**Comment by agatti** [comment]
For what it's worth, I can reproduce this on macOS and Python 3.8 with [aiohttp](https://github.com/aio-libs/aiohttp) as a failng dependency - starting from a newly created project.  Using Poetry version 1.1.4, with just `poetry init` and `poetry add "aiohttp[speedups]"` I get the very same error. 

Just in case, my configuration looks like this:

```bash
❯ poetry config --list
cache-dir = "/Users/agatti/Library/Caches/pypoetry"
experimental.new-installer = true
installer.parallel = true
virtualenvs.create = true
virtualenvs.in-project = null
virtualenvs.path = "{cache-dir}/virtualenvs"  # /Users/agatti/Library/Caches/pypoetry/virtualenvs
```

removing `/Users/agatti/Library/Caches/pypoetry` does not seem to change anything, it seems.

**Comment by lazka** [comment]
yeah, I've also had it on another machine again. I can't figure out which dependency update "fixes" it.

**Comment by agatti** [comment]
For `aiohttp` the problem is triggered by `multidict`, according to `poetry install <package> -vvv`.  But then I had this happen on other dependencies that share nothing with `aiohttp`, with the only common bit across the board being that the error shows up when a dependency needs to fetch a sdist package.

**Comment by agatti** [comment]
This is getting weird, apparently this is blocking installation of "mainstream" dependencies like flask, pandas, psycopg2 and the like.  Did anything change on pypi side to trigger this error?

**Comment by Killerjay666** [comment]
Hi guys, running into the same issue in bitbucket pipelines:
Ubuntu 18.04
Poetry version 1.0.5

Packages to install include similar ones mentioned above; flask, pandas, aiohttp
Trying to update poetry to see if that resolves it

**Comment by grucha** [comment]
I've encountered the same problem: installing `couchbase ^3.0.7` fails exactly the same way. 

I'm using `poetry 1.1.4 `

**Comment by erronppl** [comment]
First, I use Guix OS which is atypical from other Linux distros, It includes a functional package manager which builds everything from sources. As of now, Guix builds Poetry v1.1.4 with python-pkginfo v1.4.2. I updated python-pkginfo to v1.6.1 as indicated in the poetry.lock file and then rebuilt poetry. That worked for me.

Here's a list of failing packages on my host before my fix.
```
$ poetry add -vvv --dev python-language-server
$ poetry add -vvv --dev pyls-isort
$ poetry add -vvv ujson@3.2.0
$ poetry add -vvv ujson@2.0.3
```

I noticed the error when I `git clone` Poetry and ran the tests. All the tests were passing, but the ones below.

```
tests/inspection/test_info.py::test_info_setup_complex
tests/inspection/test_info.py::test_info_setup_complex_pep517_error
tests/inspection/test_info.py::test_info_setup_complex_pep517_legacy
tests/inspection/test_info.py::test_info_setup_complex_disable_build
tests/inspection/test_info.py::test_info_setup_missing_mandatory_should_trigger_pep517[version]
tests/inspection/test_info.py::test_info_setup_missing_mandatory_should_trigger_pep517[name]
tests/inspection/test_info.py::test_info_setup_missing_mandatory_should_trigger_pep517[install_requires]
```

I hope this helps.

**Comment by finswimmer** [comment]
Very strange. I'm still not able to reproduce it :thinking:

**Comment by agatti** [comment]
Tried just now:

```
❯ poetry new bug -n
Created package bug in bug
❯ cd bug
❯ poetry add pyinstaller
Creating virtualenv bug-3G4Pdgi1-py3.8 in /Users/agatti/Library/Caches/pypoetry/virtualenvs
Using version ^4.1 for pyinstaller

Updating dependencies
Resolving dependencies... (13.9s)

  AttributeError

  'NoneType' object has no attribute '_source_type'

  at /opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/poetry/inspection/info.py:548 in from_directory
      544│                         raise
      545│
      546│                     # we discovered PkgInfo but no requirements were listed
      547│
    → 548│         info._source_type = "directory"
      549│         info._source_url = path.as_posix()
      550│
      551│         return info
      552│
```

**Comment by grucha** [comment]
I have the same error with `poetry` installed from `macports` on macOS 11.1.
Uninstalled that poetry port, installed it via install script and it works fine now.

**Comment by guybogaarts** [comment]
I can confirm that what @grucha suggested solved the problem for me as well:
curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -

**Comment by agatti** [comment]
@grucha I've been using macports as well, but unless the tarball used by macports differs from what's on the poetry releases page then what could change?  The interpreter (and therefore all system and extra dependencies added) still comes from macports and would still be the same in both cases.

**Comment by neverpanic** [comment]
@agatti I'm guessing some dependency version differs subtly. Help welcome in figuring out which one.

**Comment by neverpanic** [comment]
I think I figured it out, poetry relies on [this commit](https://bazaar.launchpad.net/~tseaver/pkginfo/trunk/revision/155.1.1) in pkginfo, i.e. at least pkginfo >= 1.5.0.

**Comment by lazka** [comment]
I can confirm that updating pkginfo from 1.4.2 to 1.5.0 fixes the issue.

**Comment by lazka** [comment]
I've created #3614

**Comment by tlc28** [comment]
Hi @lazka 

> I can confirm that updating pkginfo from 1.4.2 to 1.5.0 fixes the issue for me.

It also solves the problem for me! Thanks!

**Comment by abn** [comment]
Executing the following will reproduce the issue.
```sh
podman run --rm -i --entrypoint bash python:3.8 <<EOF
set -xe
python -m pip install -q poetry
# install latest 1.5 version
python -m pip install -q pkginfo==1.4.2
python -m poetry new issue
pushd issue
python -m poetry add pyyaml==5.3.1 || true

# try with 1.5.0
python -m pip install -q pkginfo==1.5.0
python -m poetry add pyyaml==5.3.1
EOF
```

Output:
```console
+ python -m pip install -q poetry
WARNING: You are using pip version 20.2.4; however, version 21.0.1 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
+ python -m pip install -q pkginfo==1.4.2
WARNING: You are using pip version 20.2.4; however, version 21.0.1 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
+ python -m poetry new issue
Created package issue in issue
+ pushd issue
/issue /
+ python -m poetry add pyyaml==5.3.1
Creating virtualenv issue-XfWtZh9W-py3.8 in /root/.cache/pypoetry/virtualenvs

Updating dependencies
Resolving dependencies...

  AttributeError

  'NoneType' object has no attribute '_source_type'

  at /usr/local/lib/python3.8/site-packages/poetry/inspection/info.py:548 in from_directory
      544│                         raise
      545│ 
      546│                     # we discovered PkgInfo but no requirements were listed
      547│ 
    → 548│         info._source_type = "directory"
      549│         info._source_url = path.as_posix()
      550│ 
      551│         return info
      552│ 
+ true
+ python -m pip install -q pkginfo==1.5.0
WARNING: You are using pip version 20.2.4; however, version 21.0.1 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
+ python -m poetry add pyyaml==5.3.1

Updating dependencies
Resolving dependencies...

Writing lock file

Package operations: 9 installs, 0 updates, 0 removals

  • Installing pyparsing (2.4.7)
  • Installing attrs (20.3.0)
  • Installing more-itertools (8.7.0)
  • Installing packaging (20.9)
  • Installing pluggy (0.13.1)
  • Installing py (1.10.0)
  • Installing wcwidth (0.2.5)
  • Installing pytest (5.4.3)
  • Installing pyyaml (5.3.1)
```

**Comment by gruns** [comment]
if helpful, also confirming this fix: upgrading from `pkginfo==1.4.2` to `pkginfo==1.7.0` fixes the problem

**Comment by hansharhoff** [comment]
Be aware that this bug is still present in the current release version of poetry (1.1.10).
Fix is in master branch but has not been released yet.

If you hit the issue you can manually upgrade pkginfo to latest version with e.g.

pip install --upgrade pkginfo

[Fix line in master](https://github.com/python-poetry/poetry/blame/34ca1657cbd2834d98e3bb99ed76fd2df2e6b3b2/pyproject.toml#L41)
